File: raspberry-pi-ios-client/RaspberryPiClient/Data/Infrastructure/Ros/Topic/RosTopicError.swift
//
//  RosTopicError.swift
//  RaspberryPiClient
//
//  Created by Shusuke Ota on 2025/11/3.
//

import Foundation

enum RosTopicError: Error {
    case alreadySubscribed
    case failedReceiveMessage(reason: Error)
    case failedConvertStringToData
    case failedDecodeMessageToRosPublish(reason: Error)

    var description: String {
        switch self {
        case .alreadySubscribed:
            return "すでに購読済みのトピックです"
        case .failedReceiveMessage(let reason):
            return "メッセージの受信に失敗しました\(reason.localizedDescription)"
        case .failedConvertStringToData:
            return "メッセージをデータ型に変換できませんでした"
        case .failedDecodeMessageToRosPublish(let reason):
            return "メッセージの変換に失敗しました\(reason.localizedDescription)"
        }
    }
}

File: raspberry-pi-android-client/app/src/main/java/com/shusuke/raspberry_pi_client/data/infrastructure/ros/topic/RosTopicError.kt
package com.shusuke.raspberry_pi_client.data.infrastructure.ros.topic

/**
 * ROS トピック関連のエラー
 */
sealed class RosTopicError(
    override val message: String,
) : Exception(message) {

    /** すでに購読済みのトピック */
    data object AlreadySubscribed : RosTopicError("すでに購読済みのトピックです")

    /** トピックが advertise されていない */
    data object ChannelNotFound : RosTopicError("トピックが advertise されていません")

    /** メッセージの受信に失敗 */
    data class FailedReceiveMessage(val reason: Throwable) : RosTopicError(
        "メッセージの受信に失敗しました: ${reason.message ?: reason.toString()}",
    )

    /** メッセージをデータ型に変換できなかった */
    data object FailedConvertStringToData : RosTopicError("メッセージをデータ型に変換できませんでした")

    /** メッセージの変換に失敗 */
    data class FailedDecodeMessage(val reason: Throwable) : RosTopicError(
        "メッセージの変換に失敗しました: ${reason.message ?: reason.toString()}",
    )
}
